docs/proposals/diamond-design.md

```diff
@@ -0,0 +1,380 @@
+a diamond carried out by a single client _is_ an incremental upload.
+
+Using the unique ID helper command, then reusing it as a diamond reference for further uploads with `datamon diamond add` in effect implements the incremental upload feature.
+
+### Handling concurrency
+
+Premises:
+* Several diamonds can run concurrently on the same repo, resulting in distinct bundles
+* Several `add` commands may run concurrently
+* Some `split add` commands may fail, some may complete, some may be subject to a retry (see below [§ Cancellations](#handling-failures-retries-and-cancellations))
+* Only one eventual commit operation may be carried out. Any attempt to compete on committing the splits must be rejected.
+
+Design:
+* The uniqueness of the diamond ID allows for several diamonds to run in parallel. Note that uniqueness is required only for any single repo.
+* Each `add` runs with its uniquely generated internal ID: metadata can be safely altered in this namespace
+* The diamond is in effect created with the first split associated to it: there is no need to pre-propulate anything in the metadata.
+* The diamond is in effect concluded with the first commit associated to it.
+* Commit is preceded by the addition of a lock object which is removed upon completion or (controlled) failure.
+  Any attempt to commit while another commit is in progress is rejected (the operation should normally be fast, since only metadata is affected).
+  ```
+  /splits/{repo}/{diamond-id}/.committing
+  ```
+* Since there is no actual storage operation on directories, creating concurrently 2 split objects which share a common path is not an issue.
+* Lock objects might be left in some unclean state (e.g. job is cancelled): we set a timeout on locks (e.g. 60s). After that, the lock is considered stale and is ignored.
+
+### Handling conflicts
+
+Premises:
+* The bundle has a single namespace for files
+* If several nodes in the diamond upload files with same path but different data, a conflict is detected at commit time
+* Identical files do not trigger conflicts
+* The current arbitration rule is latest write wins (use nanosec timestamp here, not ksuid)
+* The loosing candidates are kept for possible future reuse (e.g. fixing merge conflicts)
+
+Design:
+* We selected latest-wins so incremental uploads which re-write several versions of same file end up with a safe deconfliction
+* The other files including the winner will be stored in a special path with the bundle metadata.
+  ```
+  /bundles/{repo}/{bundle-id}/conflicts/conflict-files-{index}.yaml
+  ```
+* The metadata to describe conflicts encompass:
+  * the reference to the originating diamond and split identifiers and timings
+  * for each conflicting path, all references (hash) to the conflicting versions (including the winner)
+
+### Handling failures, retries and cancellations
+
+Premises:
+* Some `add` operation might fail: the client has the opportunity to either _retry_ or _cancel_ the whole operation
```

Comment:
recall from discussion:  some confusion (likely on my part) about why `split add` needs _retry_ and _cancel_ notions in ui when `bundle upload` doesn't

Reply:
Will add some precision here. 
Cancellation is about the COMMIT operation, for the whole diamond, not individual split.
Retry is not a change in UI, but a change in user behaviour. I need to be more accurate/clarify these notions that are intrinsic to the new use case

Reply:
additional detail

i suppose the ui for `split add` and `bundle upload` should be as similar as possible, because users are already familiar with – or need to be familiar with, anyway, `bundle upload`, as well as diamond bundles’ `commit` command.
the only failure mode that could distinuish `add` from `upload`, as far as i can tell, is duplicate split id.  there’s also non-exist diamond id, except behavior in case of this error is, i suppose, the same failure mode as non-exist repo name for `upload` – and for `add`, for that matter.
now, i would like avoid any discrepancy between `add` and `upload` at all by avoiding `add` erroring on duplicate split ids. instead, the most recently added split id that successfully completes (about the same thing, from the user’s perspective, as completing a bundle upload successfully) is the only split id added to the bundle (in absence of any additional conflict resolution mechanisms as described elsewhere in the document).

Reply:
I've mentioned in the PR main body that seeking similarity (esp. regarding flags) between these commands _is_ my goal. I am hoping that the recap at the bottom of the doc demonstrates this point.

I am sure about what "failure mode" means to you.

I am attempting here to detail the behavior in the case of failure _during the execution of the operation_, whatever their cause.
I am making a distinction between "controlled failures" (i.e. an exception that may be caught and managed, like a network failure for instance) and "uncontrolled failures" (i.e. pod crash).

So this is definitely not about errors caused by preliminary checks (like some repo does not exist, some id is a duplicate etc).

Eventually, I removed from this spec the capability given to users to inject the split ID. This remains a debatable matter (we discussed that in part yesterday already btw).

Reply:
further discussion reveals i was operating under the false assumption that this wouldn't include a change to `diamond.md`, disregard these comments.  i'm okay with the idea of making `split add` different in some ways from `bundle upload` and in fact suppose this is rather a proposal to properly (which i define as "according to users' expectations") resolve the incremental upload functionality request as well as diamond workflow support for PT.